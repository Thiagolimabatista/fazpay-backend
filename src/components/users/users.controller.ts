import { Controller, Get, Post, Body, Patch, Param } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { UsersService } from './users.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';

@ApiTags('Users')
@Controller('users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Post()
  create(@Body() createUserDto: CreateUserDto) {
    return this.usersService.create(createUserDto);
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.usersService.findOne(id);
  }

  @Patch(':id')
  update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto) {
    return this.usersService.update(id, updateUserDto);
  }
}
