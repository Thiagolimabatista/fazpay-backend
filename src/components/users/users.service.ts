import { Injectable, BadRequestException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from '@entities/user.entity';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';

@Injectable()
export class UsersService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async create(createUserDto: CreateUserDto): Promise<User> {
    const existingEmail = await this.userRepository.findOne({
      where: { email: createUserDto.email },
    });
    if (existingEmail) {
      throw new BadRequestException('Este email já está cadastrado');
    }

    const user = this.userRepository.create({ ...createUserDto });
    return await this.userRepository.save(user);
  }

  async findOne(id: string): Promise<User> {
    const user = await this.userRepository.findOne({
      where: { id: id },
    });
    return user;
  }

  async update(id: string, updateUserDto: UpdateUserDto): Promise<User> {
    await this.userRepository.update(id, updateUserDto);
    return this.findOne(id);
  }
}
