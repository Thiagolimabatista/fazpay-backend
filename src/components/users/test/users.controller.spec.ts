import { Test, TestingModule } from '@nestjs/testing';
import { UsersController } from '../users.controller';
import { UsersService } from '../users.service';
import { User } from '@entities/user.entity';

class UsersServiceMock {
  findOne(id: string): Promise<User> {
    if (id === 'existingUserId') {
      return Promise.resolve({
        id,
        firstName: 'Thiago',
        lastName: 'FazPay',
      } as User);
    } else {
      return Promise.resolve(undefined);
    }
  }
}

describe('UsersController', () => {
  let controller: UsersController;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  let userService: UsersService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [UsersController],
      providers: [
        {
          provide: UsersService,
          useClass: UsersServiceMock,
        },
      ],
    }).compile();

    controller = module.get<UsersController>(UsersController);
    userService = module.get<UsersService>(UsersService);
  });

  describe('findOne', () => {
    it('should return an existing user by ID', async () => {
      const result = await controller.findOne('existingUserId');

      expect(result).toEqual({
        id: 'existingUserId',
        firstName: 'Thiago',
        lastName: 'FazPay',
      });
    });

    it('should return undefined for a non-existing user', async () => {
      const result = await controller.findOne('nonExistingUserId');
      expect(result).toBeUndefined();
      expect(result).toMatchSnapshot();
    });
  });
});
