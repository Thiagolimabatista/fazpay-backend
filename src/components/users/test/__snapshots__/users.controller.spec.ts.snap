// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersController findOne should return undefined for a non-existing user 1`] = `undefined`;
